src/NHibernate/Linq/Visitors/QueryModelVisitor.cs

```diff
@@ -128,16 +133,148 @@ static QueryModelVisitor()
 			ResultOperatorMap.Add<CastResultOperator, ProcessCast>();
 		}
 
-		private QueryModelVisitor(VisitorParameters visitorParameters, bool root, QueryModel queryModel)
+		private QueryModelVisitor(VisitorParameters visitorParameters, bool root, QueryModel queryModel,
+			NhLinqExpressionReturnType? rootReturnType)
 		{
 			VisitorParameters = visitorParameters;
 			Model = queryModel;
+			_rootReturnType = root ? rootReturnType : null;
 			_hqlTree = new IntermediateHqlTree(root);
 		}
 
 		private void Visit()
 		{
 			VisitQueryModel(Model);
+			AddAdditionalPostExecuteTransformer();
+		}
+
+		private void AddAdditionalPostExecuteTransformer()
+		{
+			if (_rootReturnType == NhLinqExpressionReturnType.Scalar && Model.ResultTypeOverride != null)
+			{
+				// NH-3850: handle polymorphic scalar results aggregation
+				switch ((NhExpressionType)Model.SelectClause.Selector.NodeType)
+				{
+					case NhExpressionType.Average:
+						// Polymorphic case complex to handle and not implemented. (HQL query must be reshaped for adding
+						// additional data to allow a meaningful overall average computation.)
+						// Leaving it untouched for allowing non polymorphic cases to work.
+						break;
+					case NhExpressionType.Max:
+						AddPostExecuteTransformerForResultAggregate(ReflectionCache.QueryableMethods.MaxDefinition);
+						break;
+					case NhExpressionType.Min:
+						AddPostExecuteTransformerForResultAggregate(ReflectionCache.QueryableMethods.MinDefinition);
+						break;
+					case NhExpressionType.Count:
+						AddPostExecuteTransformerForCount();
+						break;
+					case NhExpressionType.Sum:
+						AddPostExecuteTransformerForSum();
+						break;
+				}
+			}
+		}
+
+		private void AddPostExecuteTransformerForResultAggregate(MethodInfo aggregateMethodTemplate)
+		{
+			var elementType = Model.ResultTypeOverride;
+
+			LambdaExpression aggregateLambda;
+
+			if (elementType.IsNullableOrReference())
+			{
+				var inputListType = typeof(IQueryable<>).MakeGenericType(elementType);
```

Comment:
>I really do not understand what were the design rationale behind the way this post execute transformer currently works

I think it's just a copy of ListTransformer implementation, which indeed chainable.